server/src/internalClusterTest/java/org/elasticsearch/snapshots/CustomMetadataSnapshotIT.java

```diff
@@ -43,111 +41,110 @@
+        logger.info("add custom gateway metadata");
+        updateClusterState(currentState -> currentState.copyAndUpdateMetadata(metadataBuilder -> {
+            metadataBuilder.putCustom(GatewayMetadata.TYPE, new GatewayMetadata("before_restart_s_gw"));
+            metadataBuilder.putCustom(ApiMetadata.TYPE, new ApiMetadata("before_restart_ns"));
+        }));
+
+        logger.info("restart all nodes");
         internalCluster().fullRestart();
         ensureYellow();
 
-        logger.info("--> check that gateway-persistent custom metadata survived full cluster restart");
-        clusterState = clusterAdmin().prepareState().get().getState();
-        logger.info("Cluster state: {}", clusterState);
-        metadata = clusterState.getMetadata();
-        assertThat(metadata.custom(SnapshottableMetadata.TYPE), nullValue());
-        assertThat(metadata.custom(NonSnapshottableMetadata.TYPE), nullValue());
-        assertThat(
-            ((SnapshottableGatewayMetadata) metadata.custom(SnapshottableGatewayMetadata.TYPE)).getData(),
-            equalTo("before_snapshot_s_gw")
-        );
-        assertThat(
-            ((NonSnapshottableGatewayMetadata) metadata.custom(NonSnapshottableGatewayMetadata.TYPE)).getData(),
-            equalTo("after_snapshot_ns_gw")
-        );
-        // Shouldn't be returned as part of API response
-        assertThat(metadata.custom(SnapshotableGatewayNoApiMetadata.TYPE), nullValue());
-        // But should still be in state
-        metadata = internalCluster().getInstance(ClusterService.class).state().metadata();
-        assertThat(
-            ((SnapshotableGatewayNoApiMetadata) metadata.custom(SnapshotableGatewayNoApiMetadata.TYPE)).getData(),
-            equalTo("before_snapshot_s_gw_noapi")
-        );
+        var metadata = clusterAdmin().prepareState().get().getState().getMetadata();
+        logger.info("check that gateway custom metadata [{}] survived full cluster restart", metadata);
+        assertThat(metadata.<GatewayMetadata>custom(GatewayMetadata.TYPE).getData(), equalTo("before_restart_s_gw"));
+        assertThat(metadata.<ApiMetadata>custom(ApiMetadata.TYPE), nullValue());
+    }
+
+    public void testShouldExposeApiMetadata() throws Exception {
+        logger.info("add custom api metadata");
+        updateClusterState(currentState -> currentState.copyAndUpdateMetadata(metadataBuilder -> {
+            metadataBuilder.putCustom(ApiMetadata.TYPE, new ApiMetadata("before_restart_s_gw"));
+            metadataBuilder.putCustom(NonApiMetadata.TYPE, new NonApiMetadata("before_restart_ns"));
+        }));
+
+        var metadata = clusterAdmin().prepareState().get().getState().getMetadata();
+        logger.info("check that api custom metadata [{}] is visible via api", metadata);
+        assertThat(metadata.<ApiMetadata>custom(ApiMetadata.TYPE).getData(), equalTo("before_restart_s_gw"));
+        assertThat(metadata.<NonApiMetadata>custom(NonApiMetadata.TYPE), nullValue());
```

Comment:
Since this checks also gateway and api may be the test class should be renamed to `CustomMetadataScopeIT` or `CustomMetadataXContextIT`?

Reply:
Just a stupid question, why do we need the `<NonApiMetadata>` here now?

Not sure about the renaming, maybe `CustomMetadataContextIT` would be more appropriate to point so we cover the word `Context`. But I'm indifferent tbh. as far as I'm concerned the current name is good enough too :)

Reply:
>Just a stupid question, why do we need the <NonApiMetadata> here now?

This is needed to avoid explicit cast that we had here before:
```
assertThat(((SnapshottableMetadata) metadata.custom(SnapshottableMetadata.TYPE)).getData(), equalTo("before_snapshot_s"));
```

>Not sure about the renaming, maybe CustomMetadataContextIT would be more appropriate to point so we cover the word Context. But I'm indifferent tbh. as far as I'm concerned the current name is good enough too :)

Then I would rename to `CustomMetadataContextIT` as half of the test cases here are not related to the snapshots.